### ISSUE TITLE
Incorrect version parsing in `parse_sdist_filename`

### ISSUE THREAD

**Issue by woodruffw** [pypa/packaging#527]
Hi there! First of all, thanks for continuing to maintain this package -- it's extremely useful 🙂 

I'm one of the maintainers of [`pip-audit`](https://github.com/trailofbits/pip-audit), and we had a user report some strange dependency resolution behavior: https://github.com/trailofbits/pip-audit/issues/248

We were able to root-cause the bug down to a release of `cffi` (`1.0.2-2`) that uses the [implicit post releases](https://peps.python.org/pep-0440/#implicit-post-releases) syntax for specifying the post-release number, rather than the canonicalized `postN` format. This release of `cffi` is published on PyPI [here](https://pypi.org/simple/cffi/), without canonicalization, so it's likely that it was uploaded before PyPI began normalizing versions.

Because `1.0.2-2` contains a dash, the following body of `packaging.utils.parse_sdist_filename` contains an incorrect assumption and parses the source distribution name incorrectly:

```python
def parse_sdist_filename(filename: str) -> Tuple[NormalizedName, Version]:
    if filename.endswith(".tar.gz"):
        file_stem = filename[: -len(".tar.gz")]
    elif filename.endswith(".zip"):
        file_stem = filename[: -len(".zip")]
    else:
        raise InvalidSdistFilename(
            f"Invalid sdist filename (extension must be '.tar.gz' or '.zip'):"
            f" {filename}"
        )

    # We are requiring a PEP 440 version, which cannot contain dashes,
    # so we split on the last dash.
    name_part, sep, version_part = file_stem.rpartition("-")
    if not sep:
        raise InvalidSdistFilename(f"Invalid sdist filename: {filename}")

    name = canonicalize_name(name_part)
    version = Version(version_part)
    return (name, version)
```

yielding:

```python
>>> from packaging.utils import parse_sdist_filename
>>> parse_sdist_filename("cffi-1.0.2-2.tar.gz")
('cffi-1-0-2', <Version('2')>)
```

whereas we expected:

```python
>>> from packaging.utils import parse_sdist_filename
>>> parse_sdist_filename("cffi-1.0.2-2.tar.gz")
('cffi', <Version('1.0.2.post2')>)
```

TL;DR: `parse_sdist_filename` shouldn't rely on the last dash as a separator between the distribution name and the version, since PEP 440 allows dashes in non-normalized versions. Parsing this correctly poses a bit of a challenge, since distribution names can *also* contain dashes and numbers and might even contain them in pathological ways, such as:

```
# package foo3, version 1.0.0.post1
foo3-1.0.0-1.tar.gz

# package foo-3, version 1.0.0.post1
foo-3-1.0.0-1.tar.gz

# package 3_3, version 1.0.0.post1
# i'm not sure this one is valid, but i can't find a countervailing spec in any of the packaging PEPs
3-3-1.0.0-1.tar.gz
```

**Comment by woodruffw** [comment]
Here's the valid distribution regexp in [PEP 508](https://peps.python.org/pep-0508/#names):

> `^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])$`

So `3-3` and `3_3` are valid distribution names, making `3-3-1.0.0-1.tar.gz` and `3_3-1.0.0-1.tar.gz` valid pre-normalized sdist filenames.

**Comment by woodruffw** [comment]
> One thing we _could_ do here is to allow the caller to optionally provide the distribution name they were expecting to be present, so this function could either use this as a hint during parsing, or just raise an exception when what it found isn't what was expected. I'm not sure how useful this would be broadly, though.

This seems like a reasonable workaround to me, but +1 on the general usefulness concern (it's definitely useful for our needs on `pip-audit`, but I'm not sure about anyone else's needs).

**Comment by brettcannon** [comment]
I have opened https://github.com/pypa/packaging.python.org/pull/1066 to strengthen the spec around sdist file names. That implicitly updates our docs since we reference it for what the function can parse.

**Comment by uranusjr** [comment]
For some additional information. pip is able to install the package because the Simple API also provides the project name in the URL (for cffi, the endpoint is `https://pypi.org/simple/cffi/`), so pip is able to “know” that the project name is `cffi` instead of `cffi-1-0-2` and use that information to assist parsing. Perhaps `parse_sdist_filename` could do the same by accepting an optional second argument? (And if `packaging` does not, `pip-audit` can still add additional logic around the function to handle edge cases for compatibility.)

We should probably loop some of the cffi maintainers to inform them about this.

**Comment by woodruffw** [comment]
> `pip-audit` can still add additional logic around the function to handle edge cases for compatibility.)

Yep, that's what we ended up doing: https://github.com/trailofbits/pip-audit/pull/249

That suggestion otherwise sounds similar to what @di suggested in https://github.com/pypa/packaging/issues/527#issuecomment-1080880967, which sounds good to me! I'm happy to make those changes in a PR if other consumers would find them useful :slightly_smiling_face:

**Comment by woodruffw** [comment]
Updated the title to emphasize that this is an ambiguity, not a spec violation per se.

**Comment by woodruffw** [comment]
Related: the ambiguous parse here could also probably be avoided by enforcing the current guidance on sdist filenames, which suggests that distribution names be normalized from `-` to `_`. I opened https://github.com/pypa/packaging.python.org/issues/1077 as part of that, since PyPI and other hosts don't currently perform that normalization.

**Comment by uranusjr** [comment]
Well… https://peps.python.org/pep-0625/

**Comment by woodruffw** [comment]
> Well… https://peps.python.org/pep-0625/

Yep! I was just reading that PEP 🙂. Right now it says that distributions are normalized according to 503, which defers to 426, which means that they can still contain `-`. Unless I'm misunderstanding, we'd probably need a slightly stronger normalization to avoid the ambiguous parse in this case.

(But also NB that it's not currently an issue, since all recent sdists are being published with normalized versions. This would strictly be a "cherry on top" in terms of reducing ambiguity.)

**Comment by dstufft** [comment]
You missed a part:

> The name of an sdist should be {distribution}-{version}.sdist.
>
> * distribution is the name of the distribution as defined in [PEP 345](https://peps.python.org/pep-0345), and normalised according to [PEP 503](https://peps.python.org/pep-0503), e.g. 'pip', 'flit-core'.
> * version is the version of the distribution as defined in [PEP 440](https://peps.python.org/pep-0440), e.g. 20.2.
>
> Each component is escaped according to the same rules as [PEP 427](https://peps.python.org/pep-0427).

>


Specifically:


> Each component is escaped according to the same rules as [PEP 427](https://peps.python.org/pep-0427).


So the names get normalized as per PEP 503, which will convert any runs of non alphanum characters into a single ``-``, then PEP 427 escaping ``-`` into ``_``.

**Comment by uranusjr** [comment]
No because

> Each component is escaped according to the same rules as [PEP 427](https://peps.python.org/pep-0427).

In other words, like the wheel, dashes in `distribution` must be replaced by underscores.

**Comment by dstufft** [comment]
TBH, PEP 503 normalization really only means the name gets lowercased, since PEP 427 implements the exact same (effectively anyways) normalization as PEP 503 does, just uses ``_`` instead of ``-``, other than PEP 503 adds lower casing as well.

**Comment by woodruffw** [comment]
Whoops, I also missed 427. 

Just to clarify: 

> So the names get normalized as per PEP 503, which will convert any runs of non alphanum characters into a single `-`, then PEP 427 escaping `-` into `_`.

This only applies to wheels, not sdists, correct? Experimentally PyPI does not currently normalize `-` into `_` for sdist distribution names. Example: https://pypi.org/simple/pip-audit/

(That's where the ambiguity currently appears to me. The current PyPA spec says that that normalization does occur for sdists, whereas it doesn't appear to.)

**Comment by uranusjr** [comment]
PEP 427 only applies to wheels, and PEP 625 says “we’ll use the same rules for sdist”.

**Comment by woodruffw** [comment]
That's this language, right?

> Each component is escaped according to the same rules as [PEP 427](https://peps.python.org/pep-0427).

If so, that would mean a format like this, correct?

```
foo_bar-1.2.3.sdist
```

If so, that all makes sense so me, and that would be a good improvement 🙂 

(This is a significant tangent at this point, but I wonder if we couldn't avoid all of this ad-hoc parsing entirely by updating PEP 503 to include more `data-...` attributes. This could be done both both sdists and wheels: have attributes like `data-distribution=...`, `data-version=...`, and so forth.)

**Comment by uranusjr** [comment]
_Hopefully_ all these would go away when we finally standardise the JSON API. Regarding the PEP… happy to hear it all make sense for you, unfortunately not many people are entirely in agreement with the proposal (read the Discourse thread linked in the PEP).

**Comment by woodruffw** [comment]
> Would a `utils.escape_name()` function be useful in this case? I assume it would just be `utils.canonicalize_name(name).replace("-", "_")`, but at least that's easier than saying "do PEP 503 with PEP 427". 😉

I think it'd be useful for 625, although the current sdist filename format is probably foregone (since PyPI and other hosts don't do that escaping and lots of packages are already hosted with `-` in the filename).

**Comment by brettcannon** [comment]
> the current sdist filename format is probably foregone (since PyPI and other hosts don't do that escaping and lots of packages are already hosted with `-` in the filename).

For old sdists, yes, but there's nothing saying that naming can't be enforced going forward to slowly update the community through attrition.

**Comment by brettcannon** [comment]
I opened https://github.com/pypa/packaging/issues/542 for `escape_name()`.

**Comment by woodruffw** [comment]
Yep, I can open that ticket in a bit.

**Comment by dstufft** [comment]
FWIW, I was reading the [write up about this](https://blog.yossarian.net/2022/05/09/A-most-vexing-parse-but-for-Python-packaging), and FWIW I do in fact think that this is a bug with the current ``parse_sdist_filename``.

Unfortunately ``parse_sdist_filename`` has made the assumption that you can parse a sdist filename without any additional context, which absent PEP 625 is *not* true. Without something like PEP 625, we have never required sdists normalize or escape  any part of their name.

The [sdist page on packaging.python.org](https://packaging.python.org/en/latest/specifications/source-distribution-format/) suggests that canonicalizing the name and version is a de facto standard, which does not appear to actually be true? Looking through a number of projects, I don't see hardly any canonicalizing really happening. 

While I agree that we should be, at a minimum, escaping ``-`` to ``_`` in sdists, there is no actual standard or practice currently that requires it. I believe that this util function has gone unnoticed because I don't think that many projects utilize post releases, and pip itself doesn't use this function so it's not had enough use to have someone trigger the edge case until now.

Unfortunately I don't have a great answer here, ``parse_sdist_filename`` inherently can't work with all currently valid sdist names unless you break compatability to *require* passing in the additional context of the package name in question. So without breaking compatability, it is a fundamentally broken API. Ideally I guess PEP 625 gets accepted and then at least this function could be documented as only support PEP 625 compatabile file names. 

Having ``parse_sdist_filename`` exist in this library is kind of weird in general, since the libraries stated goals are:

> This library provides utilities that implement the interoperability specifications which have clearly one correct behaviour (eg: [PEP 440](http://www.python.org/dev/peps/pep-0440)) or benefit greatly from having a single shared implementation (eg: [PEP 425](http://www.python.org/dev/peps/pep-0425)).

Which doesn't actually exist for sdist filenames.

**Comment by brettcannon** [comment]
> Unfortunately `parse_sdist_filename` has made the assumption that you can parse a sdist filename without any additional context, which absent PEP 625 is _not_ true.

I think you're right, so we should probably update the docs to say that it implements PEP 625 even though it's not accepted.



> Having `parse_sdist_filename` exist in this library is kind of weird in general

I think when we accepted the function we didn't realize how many nasty edge cases that are all technically legal we had missed, or that they are in any way widespread. Maybe this is more motivation to get PEP 625 accepted. 😁

**Comment by dstufft** [comment]
Yea, it happens. packaging is full of them :P (though increasingly less of them!)

I think documenting that it implements PEP 625, possibly with warning that PEP 625 isn't accepted or enforced anywhere yet, so that there are edge cases where it will not correctly parse otherwise currently valid names to warn people is probably a reasonable option.

It's also probably the best option for what to do, since it's somewhat silly to just egregiously break compatibility for an edge case, particularly one that I think is somewhat hard to trigger in practice using modern tooling (IIRC setuptools normalizes the version by default anyways, and the cffi version that actually broke this, isn't a valid package for other reasons).

**Comment by dstufft** [comment]
Oh, I guess to technically support PEP 625, it should be implemented to also support the ``.sdist`` extension that PEP 625 implements, which is probably fine?

**Comment by dstufft** [comment]
Or maybe just document it requires the naming rules from PEP 625, but not the extension.

**Comment by brettcannon** [comment]
> Or maybe just document it requires the naming rules from PEP 625, but not the extension.

Probably this since adding the file extension is easy enough later in case that doesn't survive the PEP acceptance process.

**Comment by brettcannon** [comment]
FYI https://peps.python.org/pep-0625/ has been accepted!

**Comment by pradyunsg** [comment]
Do we want this function to serve as a validating parser for PEP 625?

**Comment by woodruffw** [comment]
I find myself looking at this again, but in an entirely different (non-`pip-audit`) context: https://github.com/pypa/pip/issues/13677 🙂 

Given that `parse_wheel_filename` *is* validating to some degree, I'd like to propose that we do a similar thing here (specifically, validate that the sdist filename is exactly `{name}-{version}.tar.gz`, where `{name}` and `{version}` are both appropriately normalized. Some ideas for that:

1. Add a `strict: bool = False` kwarg to `parse_sdist_filename`. When a user does `strict=True`, we perform the validating parse above.
    - Pros: Simple, backwards compatible. Very little new API surface.
    - Cons: Maybe _too_ backwards compatible, in the sense that people won't bother to switch to the properly validating usage?
1. Add a *new* `parse_strict_sdist_filename` (or similar, please bikeshed). The current `parse_sdist_filename` remains as is.
    - Pros: Backwards compatible.
    - Cons: Probably confusing? Also essentially duplicates the API surface, and has the same lack of an incentive to switch as option 1.
1. Turn `parse_sdist_filename` into a strict PEP 625 API, and add a `parse_legacy_sdist_filename` (or `legacy.parse_sdist_filename`?) for the original permissive behavior.
    - Pros: Does the right thing by default, but gives users the ability to opt back into the old behavior.
    - Cons: Significant breaking change.
1. Option (3), but put the "legacy" API in [packaging-legacy](https://pypi.org/project/packaging-legacy/) instead (if @di will have it 😉).
    - Pros: Same as (3), but less API surface for `packaging`.
    - Cons: Same as (3).

Of these, I personally like (3) and (4) the best, but they're also probably the most disruptive. Still, I think that disruption might be _worth it_, given that PyPI itself now enforces PEP 625.

Curious what people think about the above 🙂

**Comment by di** [comment]
@woodruffw I have been working on a new API for filenames, would love your thoughts/contributions: https://github.com/pypa/packaging/compare/main...di:packaging:filename-api (currently a year stale as I have been fixing related PyPI issues & deprecations as I attempt to replace PyPI's internal filename validation with it). It sort of does (1) and (2), giving users an easier migration path from non-strict to strict parsing and not introducing breaking changes to the existing API.

I'm fine with (4) if/when the functionality is actually deprecated or dropped from this library (which I think we should do eventually).

**Comment by woodruffw** [comment]
@di Oh, awesome! That API looks great. I'll take a closer look at it today + propose some patches via PRs 🙂

**Comment by notatallshaw** [comment]
@di looks great to me. One quick thing, I would bring up in a review is performance.

Henry and I have been spending a lot of time on improving the performance of Versions and Specifiers as they are called a lot during a resolution, a lot of this work has been doing as little as possible during initialization beyond validation.

I see in pip during a long resolution that parse_wheel_filename is called hundreds of thousands of times, I would move steps that require work but don't validate, like `self.name = canonicalize_name(name).replace("-", "_")`, to be a property.
